In [141]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules, fpmax
import matplotlib.pyplot as plt

In [142]:
print("QUESTION A:")

df = pd.read_csv('Online Retail.csv')

df['Description'] = df['Description'].str.strip()
df.dropna(axis=0, subset=['InvoiceNo'], inplace=True)
df['InvoiceNo'] = df['InvoiceNo'].astype('str')
df = df[~df['InvoiceNo'].str.contains('C')]

cleanup = (df.groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

def encode_values(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

encoded_values = cleanup.applymap(encode_values)
encoded_values.drop('POSTAGE', inplace=True, axis=1)

freq_items = apriori(encoded_values, min_support=0.03, use_colnames=True)
freq_items
freq_items.to_csv('yo.csv')


QUESTION A:


The frequent itemsets for the dataset with support of 3% are displayed. We can see that most itemsets have a support value of around 4%.

In [143]:
print("QUESTION B:")
rules = association_rules(freq_items, metric="confidence", min_threshold=0.3)
rules

QUESTION B:


antecedents                          consequents  \
0            (ALARM CLOCK BAKELIKE RED)         (ALARM CLOCK BAKELIKE GREEN)   
1          (ALARM CLOCK BAKELIKE GREEN)           (ALARM CLOCK BAKELIKE RED)   
2      (PINK REGENCY TEACUP AND SAUCER)    (GREEN REGENCY TEACUP AND SAUCER)   
3     (GREEN REGENCY TEACUP AND SAUCER)     (PINK REGENCY TEACUP AND SAUCER)   
4     (ROSES REGENCY TEACUP AND SAUCER)    (GREEN REGENCY TEACUP AND SAUCER)   
5     (GREEN REGENCY TEACUP AND SAUCER)    (ROSES REGENCY TEACUP AND SAUCER)   
6             (JUMBO BAG RED RETROSPOT)            (JUMBO BAG PINK POLKADOT)   
7             (JUMBO BAG PINK POLKADOT)            (JUMBO BAG RED RETROSPOT)   
8   (JUMBO SHOPPER VINTAGE RED PAISLEY)            (JUMBO BAG RED RETROSPOT)   
9             (JUMBO BAG RED RETROSPOT)  (JUMBO SHOPPER VINTAGE RED PAISLEY)   
10             (JUMBO STORAGE BAG SUKI)            (JUMBO BAG RED RETROSPOT)   
11            (JUMBO BAG RED RETROSPOT)             (JUMBO STORAGE BAG SUKI)   
12            (LUNCH BAG  BLACK SKULL.)            (LUNCH BAG RED RETROSPOT)   
13            (LUNCH BAG RED RETROSPOT)            (LUNCH BAG  BLACK SKULL.)   

    antecedent support  consequent support   support  confidence       lift  \
0             0.050995            0.047550  0.031053    0.608944  12.806462   
1             0.047550            0.050995  0.031053    0.653061  12.806462   
2             0.037166            0.049248  0.030713    0.826371  16.779804   
3             0.049248            0.037166  0.030713    0.623645  16.779804   
4             0.051722            0.049248  0.037263    0.720450  14.629045   
5             0.049248            0.051722  0.037263    0.756650  14.629045   
6             0.101504            0.059098  0.040029    0.394359   6.673028   
7             0.059098            0.101504  0.040029    0.677340   6.673028   
8             0.057011            0.101504  0.032994    0.578723   5.701477   
9             0.101504            0.057011  0.032994    0.325048   5.701477   
10            0.057448            0.101504  0.035129    0.611486   6.024253   
11            0.101504            0.057448  0.035129    0.346080   6.024253   
12            0.061766            0.075885  0.031101    0.503535   6.635457   
13            0.075885            0.061766  0.031101    0.409847   6.635457   

    leverage  conviction  
0   0.028628    2.435585  
1   0.028628    2.735368  
2   0.028883    5.475760  
3   0.028883    2.558314  
4   0.034716    3.401012  
5   0.034716    3.896768  
6   0.034030    1.553566  
7   0.034030    2.784651  
8   0.027207    2.132793  
9   0.027207    1.397119  
10  0.029297    2.312650  
11  0.029297    1.441388  
12  0.026414    1.861389  
13  0.026414    1.589813

The association rules for the frequent itemsets with 30% confidence are displayed. Based on the highest lift values, items in row 2-3 and the last 4 rows have the strongest rules. Most rules have over 50% confidence which is great.

In [144]:
print("QUESTION C:")
maxItems = fpmax(encoded_values, min_support=0.03,use_colnames=True).sort_values(by=['support'],ascending=False)
maxItems

QUESTION C:


support                               itemsets
113  0.109656   (WHITE HANGING HEART T-LIGHT HOLDER)
112  0.096507             (REGENCY CAKESTAND 3 TIER)
111  0.081805                        (PARTY BUNTING)
110  0.070597        (ASSORTED COLOUR BIRD ORNAMENT)
109  0.067200     (SET OF 3 CAKE TINS PANTRY DESIGN)
..        ...                                    ...
4    0.030519          (SMALL WHITE HEART OF WICKER)
3    0.030519   (SET OF 60 PANTRY DESIGN CAKE CASES)
2    0.030519             (LOVE BUILDING BLOCK WORD)
1    0.030228     (RETROSPOT HEART HOT WATER BOTTLE)
0    0.030131  (COLOUR GLASS T-LIGHT HOLDER HANGING)

[114 rows x 2 columns]

The maximum itemsets for the frequent items in (A) are displayed. This means there is 114 itemsets which are frequent but none of their supersets are frequent.

In [145]:
print("QUESTION D:")
freq_items.head(10)

QUESTION D:


support                         itemsets
0  0.046337         (6 RIBBONS RUSTIC CHARM)
1  0.040175    (60 TEATIME FAIRY CAKE CASES)
2  0.047550     (ALARM CLOCK BAKELIKE GREEN)
3  0.037991      (ALARM CLOCK BAKELIKE PINK)
4  0.050995       (ALARM CLOCK BAKELIKE RED)
5  0.033430   (ANTIQUE SILVER T-LIGHT GLASS)
6  0.070597  (ASSORTED COLOUR BIRD ORNAMENT)
7  0.045318   (BAKING SET 9 PIECE RETROSPOT)
8  0.032654            (BATHROOM METAL SIGN)
9  0.036050    (CHARLOTTE BAG PINK POLKADOT)

The top-10 frequent items with the highest support are displayed. The support starts high at almost 11% and quickly decreases to 6% for the 10th item.